chain/client-primitives/src/types.rs

```diff
@@ -32,6 +32,8 @@ use serde::Serialize;
 pub enum Error {
     #[error("Chain: {0}")]
     Chain(#[from] near_chain_primitives::Error),
+    #[error("{0}: {1}")]
+    ChainWithCtx(&'static str, #[source] near_chain_primitives::Error),
```

Comment:
> Instead, the fn head itself should make it clear that it is failing. 

That doesn't comply with "Each unique fal­lible ex­pres­sion should have at least one unique er­ror. " as it gives you a unique error per callable function, not per unique fallible expression. In particular such thing won't indicate which call failed if you call the same function more than once from a single context (you can imagine a function which calls a.head() and then b.head() for example). Not that my context strings are perfect, as they are just good enough for the current implementation, but they can be customized later if needed.

> However, when I contribute to an area I touch very occasionally, I try as much as possible to mimic the surrounding code.

Totally agree. I don't want to force my opinions in other peoples' domain. I just thought (basing on our previous discussions with @nagisa and other folks) that moving to strongly-typed errors with meaningful context was an already agreed-upon strategy. But if that's not the case, then here goes a context-free chain error :P.

Reply:
> That's not the right place to do this. We shouldn't make all the callers of chain.head() add the same "chain.head()" context. Instead, the fn head itself should make it clear that it is failing. We should add the context here: https://github.com/matklad/nearcore/blob/91756eedae1ab4fe1a276a685a8a65ca22bdf402/chain/chain/src/store.rs#L817-L824 (So, to the IO variant of chain_primitives::Error)

> That doesn't comply with "Each unique fal­lible ex­pres­sion should have at least one unique er­ror. " as it gives you a unique error per callable function, not per unique fallible expression.

Correct. I’m strongly advising for making each “unique” (clarification: semantically different) expression have a reasonably unique error associated with it. In general, having each fallible expression to be on the hook for setting this context up, makes it much easier to enforce this via code review and such.

Deserializing a configuration file at path `a` and a data file at path `b` will be quite distinct operations from the standpoint of the user reading an error message, and being able to tell which one failed at a glance will be a tremendous help in addressing any problem that involves these operations. In practice these sorts of semantically different operations will also show up in the code as “unique” expressions (even if they differ by the semantic purpose of the arguments passed into a same function), hence the wording of the rule-of-thumb.

Conversely, it isn’t super clear whether one call to `chain.head()` have meaningful semantic differences from a `chain.head()` call in a different place. If not, I can see @matklad’s argument making sense. Except in that case the general pattern breaks down somewhat. It is somewhat easier to both write and review the code if nobody needs to think whether some fallible expression should `?` propagate an error from the call, or add-a-context-and-propagate with `map_err(...)?`.

> Totally agree. I don't want to force my opinions in other peoples' domain. I just thought (basing on our previous discussions with @nagisa and other folks) that moving to strongly-typed errors with meaningful context was an already agreed-upon strategy. But if that's not the case, then here goes a context-free chain error :P.

Aha! I haven’t actually gone through the effort of getting everybody to agree with my (and wg-error-handling’s, for that matter) take on the topic. This is more of a ruleset I myself will follow in the code I touch (in absence of any other rules, anyway) and may ask people to follow where it meaningfully impacts the quality of the diagnostics or the comprehensibility of the code.

Otherwise, I would tend to agree with the idea that refactoring of the code to adhere to these rules-of-thumb should probably be independent of feature dev work, but it isn’t clear to me if this PR is actually executing any refactors per-se – most of the new `map_err`s appear to be in newly added code.

Reply:
So my summary here would be that: 

* it's indeed true that tagging context when we call `.head()` is in theory better
* but it's also true that we are not really tagging with the right context there: practically, the current diff is a worse version of tagging the context in the impl of `.head()`
* it's also arguable whether various calls to head really need different contexts. 
* it's also arguable whether this is a self-contained incremental improvement (boy scout rule), or an incomplete, partical refactor (https://xkcd.com/927/)

Ultimately, the overriding considiration for me would be that this needs some design, and that this is out of scope for PR in question

Reply:
going with bare chain::Error for now to comply with the code around.